curated_transformers/models/attention.py

```diff
@@ -61,12 +70,112 @@ def forward(
+            # negative value to zero them out during softmax normalization.
+            batch, seq_len = attn_mask.shape
+            attn_scores += attn_mask.logit_mask.view(batch, 1, 1, seq_len)
 
         attn_weights = attn_scores.softmax(dim=-1)
         attn_values = self.dropout(attn_weights @ v)
 
         return attn_values
+
+
+# https://www.tensorflow.org/text/tutorials/transformer#multi-head_attention
+class SelfAttention(Module):
+    """Transformer self-attention layer (Vaswani et al., 2016)."""
+
+    def __init__(
+        self,
+        *,
+        dropout_prob: float = 0.1,
+        hidden_width: int = 768,
+        num_attention_heads: int = 12,
+    ):
+        """Construct a self-attention layer.
+
+        dropout_prob (float): dropout to apply between the self-attention
+            and output layers (default: 0.1).
+        hidden_width (int): hidden width of the layer (default: 768).
+        num_attention_heads (int): number of attenion heads (default: 12).
+        """
+        super().__init__()
+
+        self.model_dim = hidden_width
+        self.num_heads = num_attention_heads
+        if self.model_dim % self.num_heads != 0:
+            raise ValueError(
+                Errors.E003.format(
+                    hidden_width=self.model_dim, num_heads=self.num_heads
+                )
+            )
+
+        self.dims_per_head = self.model_dim // self.num_heads
+        self.attention = ScaledDotProductAttention(dropout_prob=dropout_prob)
+        self.input = Linear(self.model_dim, self.model_dim * 3)
+        self.output = Linear(self.model_dim, self.model_dim)
+
+    def forward(self, x: Tensor, attn_mask: Optional[AttentionMask]) -> Tensor:
+        """
+        Apply self-attention layer to the input.
+
+        Sequence elements that are marked with `False` in the attention mask
+        are ignored by the attention mechanism (if a mask is provided).
+
+        x (Tensor): input to apply self-attention to.
+        attn_mask (Optional[AttentionMask]): attention mask.
+
+        Shapes:
+            x - (batch, seq_len, width)
+            attn_mask - (batch, seq_len)
+        """
+
+        proj = self.input(x)
+        q, k, v = proj.chunk(3, dim=-1)
+
+        # (batch, head, seq_len, width_per_head)
+        k = split_heads(k, self.num_heads)
+        q = split_heads(q, self.num_heads)
+        v = split_heads(v, self.num_heads)
+
+        # (batch, seq_len, width)
+        attn = combine_heads(self.attention(k, q, v, attn_mask))
+        out = self.output(attn)
+
+        return out
+
+
+def combine_heads(x: Tensor) -> Tensor:
+    """
+    Combine the attention head representations. The inverse of
+    'split_heads'.
+
+    Shapes:
+        x - (batch, head, seq_len, width_per_head)
+        output - (batch, seq_len, width)
+    """
+    batch_size, head, seq_len, model_dim = x.size()
+    return x.transpose(1, 2).contiguous().view(batch_size, seq_len, head * model_dim)
```

Comment:
This reads better if it came after `split_heads`, given that it's the inverse of the former.

Reply:
Indeed, fixed!